CONTRIBUTING.md

```diff
@@ -0,0 +1,62 @@
+    * Ensure that nobody else has already begun working on this issue. 
+    If they have, please try to contact them to collaborate
+    * If nobody has been assigned for the issue and you would like to work on it, make a comment on the issue to inform the community of your intentions to begin work. (So we can avoid duplication of efforts)
+    * We suggest using standard Github best practices for contributing: fork the repo, branch from the HEAD of develop, make some commits on your branch, and submit a PR from the branch to develop.
+    More detail on this is below
+    * Be sure to include a relevant change log entry in the Pending section of CHANGELOG.md (see file for log format)
+
+Note that for very small or clear problems (such as typos), or well isolated improvements, it is not required to an open issue to submit a PR. But be aware that for more complex problems/features touching multiple parts of the codebase, if a PR is opened before an adequate design discussion has taken place in a github issue, that PR runs a larger likelihood of being rejected.
+
+Looking for a good place to start contributing? How about checking out some good first issues
+
+## Branch Structure
+
+Libff has its default branch as `develop`, which is where PRs are merged into. The `master` branch should only contain code that is part of a release. Releases will be periodically made. All other branches should be assumed to be miscellaneous feature development branches.
+
+All downstream users should be using tagged versions of the library.
+
+## How to work on a fork
+Please skip this section if you're familiar with contributing to opensource github projects.
+
+First fork the repo from the github UI, and clone it locally.
+Then in the repo, you want to add the repo you forked from as a new remote. You do this as:
+```bash
+git remote add upstream git@github.com:scipr-lab/libiop.git
+```
+
+Then the way you make code contributions is to first think of a branch name that describes your change.
+Then do the following:
+```bash
+git checkout master
+git pull upstream master
+git checkout -b $BRANCH_NAME
+```
+and then work as normal on that branch, and PR'ing to upstream master when you're done =)
+
+## Updating documentation
+
+All PRs should aim to leave the code more documented than it started with.
+Please don't assume that its easy to infer what the code is doing, 
+as that is usually not the case for these complex protocols unless one has been working with it recently. 
+(Even if you understand the paper!)
+
+Its often very useful to describe what is the high level view of what a code block is doing,
+and either refer to the relevant section of a paper or include a short proof/argument for why it makes sense before the actual logic.
+
+## Performance improvements
+
+All performance improvements should be accompanied with benchmarks improving, or otherwise have it be clear that things have improved.
```

Comment:
Just a side note on that. Ideally some benchmarks should be done on the CI to track down the performance impact of the changes parked in a branch before it is merged to the target branch (if Github Action is used, it may simply be enough to run a suite of benchmark "on pull-request"). A lot of seemingly "harmless" code (not even aimed at improving perfs) can end up affecting the perfs quite a bit. These type of changes can include - for instance - "code cleanups", "building layers of abstraction" etc. The right tradeoffs always need to be taken between "performances" and "code clarity" (which in many case implies better "code soundness" - i.e. more chances to catch issues in a well designed and "less opaque" code -> this is just strengthening the point above about documentation :) )